### PR TITLE
feat: Solidity type parser

### PR DESCRIPTION
**Pull request by prestwich**
supersedes #180 

## Motivation
Both dyn-abi and json-abi rely on string representations of solidity types

## Solution
Break the parser logic from dyn-abi into a new crate both can inherit

- [x] todo: integrate into json-abi

## PR Checklist

- [x] Added Tests
- [x] Added Documentation
- [x] Breaking changes

### CODE REVIEW

File: crates/dyn-abi/src/lib.rs [hunk 1]
```diff
@@ -46,7 +46,7 @@ pub use value::DynSolValue;
 mod token;
 pub use token::DynToken;
 
-pub mod parser;
+pub mod resolve;
```
Comment (DaniPopes) [inline crates/dyn-abi/src/lib.rs:50]: maybe re-export the `newcrate as parser`

File: crates/dyn-abi/src/ty.rs [hunk 2]
```diff
@@ -147,7 +165,7 @@ impl DynSolType {
     /// ```
     #[inline]
     pub fn parse(s: &str) -> DynAbiResult<Self> {
-        TypeSpecifier::try_from(s).and_then(|t| t.resolve_basic_solidity())
+        TypeSpecifier::parse(s)?.resolve()
```
Comment (DaniPopes) [inline crates/dyn-abi/src/ty.rs:168]: `and_then` has better codegen, I think re-uses the first error's stack allocation or something

File: crates/sol-type-str/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,24 @@
+[package]
+name = "alloy-sol-type-str"
```
Comment (DaniPopes) [inline crates/sol-type-str/Cargo.toml:2]: wdyt of `alloy-sol-type-parser`?

File: crates/sol-type-str/src/root.rs [hunk 4]
```diff
@@ -0,0 +1,114 @@
...
+        }
+    }
+
+    /// The string underlying this type. The type name.
+    #[inline]
+    pub const fn span(self) -> &'a str {
+        self.0
+    }
+
+    /// Returns true if the type is a basic Solidity type.
```
Comment (DaniPopes) [inline crates/sol-type-str/src/root.rs:78]:
```suggestion
    /// Returns `Ok(())` if the type is a basic Solidity type.
```

File: crates/json-abi/src/internal_type.rs [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+    where
+        D: Deserializer<'de>,
+    {
+        deserializer.deserialize_str(ItVisitor).map(Into::into)
+    }
+}
+
+impl InternalType {
+    /// Parse a string into an instance, taking ownership of data
+    pub fn parse(s: &str) -> Option<Self> {
```
Comment (DaniPopes) [inline crates/json-abi/src/internal_type.rs:91]: some `#[inline]`s perhaps :flushed:

File: crates/dyn-abi/src/resolve.rs [hunk 6]
```diff
@@ -0,0 +1,351 @@
...
+
+use crate::{DynAbiError, DynAbiResult, DynSolType};
+use alloc::vec::Vec;
+
+use alloy_json_abi::{EventParam, Param};
+use alloy_sol_type_str::{
+    Error as TypeStrError, RootType, TupleSpecifier, TypeSpecifier, TypeStem,
+};
+
+pub(crate) trait ResolveSolType {
```
Comment (DaniPopes) [inline crates/dyn-abi/src/resolve.rs:13]: make public?

File: crates/json-abi/src/param.rs [hunk 7]
```diff
@@ -15,22 +22,27 @@ use serde::{Deserialize, Deserializer, Serialize, Serializer};
...
-    pub internal_type: Option<String>,
+    pub internal_type: Option<InternalType>,
 }
 
 impl fmt::Display for Param {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         if let Some(internal_type) = &self.internal_type {
-            f.write_str(internal_type)?;
+            write!(f, "{} ", internal_type)?;
             f.write_str(" ")?;
```
Comment (DaniPopes) [inline crates/json-abi/src/param.rs:46]: double space

File: crates/json-abi/src/param.rs [hunk 8]
```diff
@@ -103,26 +219,32 @@ impl Param {
...
-    pub internal_type: Option<String>,
+    pub internal_type: Option<InternalType>,
 }
 
 impl fmt::Display for EventParam {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
         if let Some(internal_type) = &self.internal_type {
-            f.write_str(internal_type)?;
+            write!(f, "{} ", internal_type)?;
             f.write_str(" ")?;
```
Comment (DaniPopes) [inline crates/json-abi/src/param.rs:248]: double space

File: crates/json-abi/src/param.rs [hunk 9]
```diff
@@ -178,20 +400,24 @@ impl EventParam {
...
+            internal_type: self.borrowed_internal_type(),
             components: Cow::Borrowed(&self.components),
         }
     }
 }
 
-#[derive(Deserialize, Serialize)]
-#[serde(bound(deserialize = "<[T] as ToOwned>::Owned: Default + Deserialize<'de>"))]
+#[derive(Deserialize, Serialize, Debug)]
+#[serde(bound(deserialize = "<[T] as ToOwned>::Owned: Default + Deserialize<'de>, 'a: 'de"))]
```
Comment (DaniPopes) [inline crates/json-abi/src/param.rs:420]: lifetime bound is added by serde by doing `#[serde(borrow)]`, i dont think you need to add it here too

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/dyn-abi/src/ty.rs (hunk 2 above)
Comment (prestwich) [reply crates/dyn-abi/src/ty.rs:168]: `and_then` cannot be used to change the result type. I'll remove the `?` another way

File: crates/sol-type-str/Cargo.toml (hunk 3 above)
Comment (prestwich) [reply crates/sol-type-str/Cargo.toml:2]: improvement.

File: crates/dyn-abi/src/resolve.rs [hunk 10]
```diff
@@ -0,0 +1,384 @@
...
+/// assert_eq!("bytes32".resolve()?, DynSolType::FixedBytes(32));
+/// # Ok(())
+/// # }
+/// ```
+pub trait ResolveSolType {
+    /// Resolve this object into a [`DynSolType`].
+    fn resolve(&self) -> DynAbiResult<DynSolType>;
+}
+
+impl ResolveSolType for &str {
```
Comment (DaniPopes) [inline crates/dyn-abi/src/resolve.rs:44]:
```suggestion
impl ResolveSolType for str {
```
plus impls for `&`, `&mut` etc?

File: crates/dyn-abi/src/resolve.rs [hunk 11]
```diff
@@ -0,0 +1,384 @@
...
+        self.types
+            .iter()
+            .map(TypeSpecifier::resolve)
+            .collect::<Result<Vec<_>, _>>()
+            .map(DynSolType::Tuple)
+    }
+}
+
+impl ResolveSolType for TypeStem<'_> {
+    /// Resolve the type string into a basic Solidity type if possible.
```
Comment (DaniPopes) [inline crates/dyn-abi/src/resolve.rs:109]: trait impls' doc comments override the trait's

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/dyn-abi/src/ty.rs (hunk 2 above)
Comment (DaniPopes) [reply crates/dyn-abi/src/ty.rs]: If they have different error then that's fine

File: crates/json-abi/src/param.rs (hunk 9 above)
Comment (prestwich) [reply crates/json-abi/src/param.rs:420]: it refused to compile without. I am not familiar enough with the serde internals to know why yet

File: crates/dyn-abi/src/ty.rs (hunk 2 above)
Comment (prestwich) [reply crates/dyn-abi/src/ty.rs]: used a map_err, which should still have better codegen than `?`

File: crates/dyn-abi/src/resolve.rs [hunk 12]
```diff
@@ -56,6 +56,34 @@ where
...
+where
+    T: ResolveSolType,
+{
+    fn resolve(&self) -> DynAbiResult<DynSolType> {
+        (**self).resolve()
+    }
+}
+
+#[cfg(feature = "std")]
+impl<T> ResolveSolType for std::sync::Arc<T>
```
Comment (DaniPopes) [inline crates/dyn-abi/src/resolve.rs:78]: this is in alloc

File: crates/json-abi/src/internal_type.rs [hunk 13]
```diff
@@ -0,0 +1,413 @@
+use core::fmt;
+
+use alloc::string::{String, ToString};
+
+use alloy_sol_type_parser::TypeSpecifier;
+use serde::{de::Visitor, Deserialize, Deserializer, Serialize, Serializer};
+
+#[derive(Clone, Debug, PartialEq, Eq, Hash)]
```
Comment (DaniPopes) [inline crates/json-abi/src/internal_type.rs:8]: nit: derives below docs

File: crates/json-abi/src/internal_type.rs [hunk 14]
```diff
@@ -0,0 +1,413 @@
...
+        }
+    }
+}
+
+impl Serialize for BorrowedInternalType<'_> {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: Serializer,
+    {
+        serializer.serialize_str(&format!("{}", self))
```
Comment (DaniPopes) [inline crates/json-abi/src/internal_type.rs:270]: `collect_str`

**Review by DaniPopes — CHANGES_REQUESTED** [review]
after this LGTM. I imagine we can't get the parser types in `InternalType` because of lifetimes

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**